File: C++/zero-array-transformation-i.cpp
// Time:  O(n + q)
// Space: O(n)

// line sweep
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        vector<int> events(size(nums) + 1);
        for (const auto& q : queries) {
            ++events[q[0]];
            --events[q[1] + 1];
        }
        for (int i = 0, curr = 0; i < size(nums); ++i) {
            curr += events[i];
            if (nums[i] > curr) {
                return false;
            }
        }
        return true;
    }
};

File: Python/zero-array-transformation-i.py
# Time:  O(n + q)
# Space: O(n)

# line sweep
class Solution(object):
    def isZeroArray(self, nums, queries):
        """
        :type nums: List[int]
        :type queries: List[List[int]]
        :rtype: bool
        """
        events = [0]*(len(nums)+1)
        for l, r in queries:
            events[l] += 1
            events[r+1] -= 1
        curr = 0
        for i in xrange(len(nums)):
            curr += events[i]
            if nums[i] > curr:
                return False
        return True
